### PR TITLE
	Added "--grant-all" parameter which allows to grant all runtime permissions during installation on Marshmallow and above devices

### PR DESCRIPTION
**Pull request by kirillzh**
Related to #355

### CODE REVIEW

File: spoon-client/src/main/java/com/squareup/spoon/Spoon.java [hunk 1]
```diff
@@ -249,7 +251,7 @@ static StackTraceElement findTestClassTraceElement(StackTraceElement[] trace) {
 
   private static StackTraceElement extractStackElement(StackTraceElement[] trace, int i) {
     //Stacktrace length changed in M
-    int testClassTraceIndex = Build.VERSION.SDK_INT >= 23 ? (i - 2) : (i - 3);
+    int testClassTraceIndex = Build.VERSION.SDK_INT >= MARSHMALLOW_API_LEVEL ? (i - 2) : (i - 3);
```
Comment (kirillzh) [inline spoon-client/src/main/java/com/squareup/spoon/Spoon.java]: Could've use `Build.VERSION_CODES` but it requires additional dependency in this module.

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 2]
```diff
@@ -314,10 +310,31 @@ public DeviceResult run(AndroidDebugBridge adb) {
     return result.build();
   }
 
+  private void addCodeCoverageInstrumentationArgs(RemoteAndroidTestRunner runner, IDevice device)
```
Comment (kirillzh) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: Code style checker complained that `run` method got too long, so I created couple of methods.

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 3]
```diff
@@ -314,10 +310,31 @@ public DeviceResult run(AndroidDebugBridge adb) {
...
+    runner.addInstrumentationArg("coverage", "true");
+    runner.addInstrumentationArg("coverageFile", coveragePath);
+  }
+
+  private void addShardingInstrumentationArgs(RemoteAndroidTestRunner runner) {
+    runner.addInstrumentationArg("numShards", Integer.toString(numShards));
+    runner.addInstrumentationArg("shardIndex", Integer.toString(shardIndex));
+  }
+
+  private void cleanResultsDirectory(File dir, File newDir, DeviceResult.Builder result)
```
Comment (kirillzh) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: I have no idea, how to name this method 😄

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java [hunk 4]
```diff
@@ -547,6 +556,11 @@ public SpoonRunner build() {
         converter = FileConverter.class) //
     public File initScript;
 
+    @Parameter(names = { "--grant-all" },
+        description = "Grant all runtime permissions during installation on Marshmallow and above"
+                + " devices") //
```
Comment (kirillzh) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java]: I hope, wording is fine.

**Comment by jainsahab** [comment]
@kirillzh , Out of curiosity, Is it something which will also allow granting **dangerous permission** during installation ?

**Comment by kirillzh** [comment]
@jainsahab, this parameter will just grant all runtime permissions during installation of application you are testing. I'm not sure what you mean by "dangerous permission".

**Comment by jainsahab** [comment]
image

@kirillzh as the android docs says dangerous permission are something that requires an app user approval at runtime. So I believe, it should work for dangerous permission too.

**Comment by kirillzh** [comment]
@jainsahab I am not sure, how this is relevant.

**Comment by kirillzh** [comment]
@jainsahab, sorry, I am not sure how this is relevant...

**Comment by edenman** [comment]
Thanks @kirillzh!  I just published this as 1.6.0, should be live in a couple hours.

**Pull request closed without merging**